File: src/main/java/cz/vhromada/catalog/web/validator/DateRangeValidator.java
package cz.vhromada.catalog.web.validator;

import java.util.regex.Pattern;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;

import cz.vhromada.catalog.web.validator.constraints.DateRange;
import cz.vhromada.common.utils.Constants;

/**
 * A class represents validator for date range constraint.
 *
 * @author Vladimir Hromada
 */
public class DateRangeValidator implements ConstraintValidator<DateRange, String> {

    /**
     * Date pattern
     */
    private static final Pattern PATTERN = Pattern.compile("\\d{4}");

    /**
     * Minimal date
     */
    private int minDate;

    /**
     * Maximal date
     */
    private int maxDate;

    @Override
    public void initialize(final DateRange dateRange) {
        minDate = dateRange.value();
        maxDate = Constants.CURRENT_YEAR;
    }

    @Override
    public boolean isValid(final String date, final ConstraintValidatorContext constraintValidatorContext) {
        if (date == null) {
            return false;
        }
        if (!PATTERN.matcher(date).matches()) {
            return false;
        }

        final int intValue = Integer.parseInt(date);
        return intValue >= minDate && intValue <= maxDate;
    }

}

File: src/main/kotlin/cz/vhromada/catalog/web/validator/DateRangeValidator.kt
package cz.vhromada.catalog.web.validator

import cz.vhromada.catalog.web.validator.constraints.DateRange
import cz.vhromada.common.utils.Constants
import java.util.regex.Pattern
import javax.validation.ConstraintValidator
import javax.validation.ConstraintValidatorContext

/**
 * A class represents validator for date range constraint.
 *
 * @author Vladimir Hromada
 */
class DateRangeValidator : ConstraintValidator<DateRange, String> {

    /**
     * Minimal date
     */
    private var minDate: Int = 0

    override fun initialize(dateRange: DateRange) {
        minDate = dateRange.value
    }

    override fun isValid(date: String?, constraintValidatorContext: ConstraintValidatorContext): Boolean {
        if (date == null || !PATTERN.matcher(date).matches()) {
            return false
        }

        return Integer.parseInt(date) in minDate..Constants.CURRENT_YEAR
    }

    @Suppress("CheckStyle")
    companion object {

        /**
         * Date pattern
         */
        private val PATTERN = Pattern.compile("\\d{4}")

    }

}
